# Questão 3 — Power Consumption of Tetouan City

## Etapa B — Python / Pandas

**Objetivo:** identificar qual das três zonas apresenta o maior pico de consumo e observar o efeito da condição de temperatura nos registros de maior demanda.




## Carregamento do dataset

In [1]:
import os
import pandas as pd

arquivo_csv = "Tetuan City power consumption.csv"

if not os.path.exists(arquivo_csv):
    from google.colab import files
    uploaded = files.upload()
    arquivo_csv = next(iter(uploaded.keys()))

dados = pd.read_csv(arquivo_csv)

print(f"Quantidade de registros: {len(dados)}")
display(dados.head())


Saving Tetuan City power consumption.csv to Tetuan City power consumption.csv
Quantidade de registros: 52416


,DateTime,Temperature,Humidity,Wind Speed,general diffuse flows,diffuse flows,Zone 1 Power Consumption,Zone 2 Power Consumption,Zone 3 Power Consumption
0,1/1/2017 0:00,6.559,73.8,0.083,0.051,0.119,34055.69620,16128.87538,20240.96386
1,1/1/2017 0:10,6.414,74.5,0.083,0.070,0.085,29814.68354,19375.07599,20131.08434
2,1/1/2017 0:20,6.313,74.5,0.080,0.062,0.100,29128.10127,19006.68693,19668.43373
3,1/1/2017 0:30,6.121,75.0,0.083,0.091,0.096,28228.86076,18361.09422,18899.27711
4,1/1/2017 0:40,5.921,75.7,0.081,0.048,0.085,27335.69620,17872.34043,18442.40964


## 1. Renomear as três variáveis de consumo

In [2]:
dados = dados.rename(columns={
    "Zone 1 Power Consumption": "Consumo_Zona_1",
    "Zone 2  Power Consumption": "Consumo_Zona_2",
    "Zone 3  Power Consumption": "Consumo_Zona_3"
})

print("Colunas após a renomeação:")
print(dados.columns.tolist())


Colunas após a renomeação:
['DateTime', 'Temperature', 'Humidity', 'Wind Speed', 'general diffuse flows', 'diffuse flows', 'Consumo_Zona_1', 'Consumo_Zona_2', 'Consumo_Zona_3']


## 2. Determinar o consumo máximo registrado em cada zona

In [3]:
maximos = dados[
    ["Consumo_Zona_1", "Consumo_Zona_2", "Consumo_Zona_3"]
].max()

print("Consumo máximo por zona:")
print(maximos)


Consumo máximo por zona:
Consumo_Zona_1    52204.39512
Consumo_Zona_2    37408.86076
Consumo_Zona_3    47598.32636
dtype: float64


### Resultado dos máximos

- **Zona 1:** 52204.39512
- **Zona 2:** 37408.86076
- **Zona 3:** 47598.32636


## 3. Identificar qual zona apresenta o maior pico de consumo

In [4]:
zona_maior = maximos.idxmax()
maximo_maior = maximos.max()

print(f"Zona com maior pico: {zona_maior}")
print(f"Maior pico de consumo: {maximo_maior:.5f}")


Zona com maior pico: Consumo_Zona_1
Maior pico de consumo: 52204.39512


### Resultado

A **Zona 1** apresenta o maior pico de consumo, com valor máximo de **52204.39512**.


## 4. Calcular 70% do valor máximo e criar o DataFrame de alto consumo

In [5]:
limiar_70 = 0.70 * maximo_maior

df_alto = dados[dados[zona_maior] > limiar_70].copy()

print(f"70% do máximo da Zona 1: {limiar_70:.5f}")
print(f"Registros acima do limiar: {len(df_alto)}")
display(df_alto.head())


70% do máximo da Zona 1: 36543.07658
Registros acima do limiar: 14718


,DateTime,Temperature,Humidity,Wind Speed,general diffuse flows,diffuse flows,Consumo_Zona_1,Consumo_Zona_2,Consumo_Zona_3
106,1/1/2017 17:40,14.93,60.83,0.077,76.500,94.000,38321.01266,23613.37386,25179.75904
107,1/1/2017 17:50,14.71,61.87,0.086,26.510,28.080,39129.11392,24167.78116,25520.96386
108,1/1/2017 18:00,14.48,63.27,0.080,18.000,18.290,39560.50633,24455.92705,25544.09639
109,1/1/2017 18:10,14.29,63.94,0.078,6.994,7.020,39517.97468,24463.22188,25619.27711
110,1/1/2017 18:20,14.09,64.87,0.078,2.818,2.763,39596.96203,24000.00000,25798.55422


### Resultado

O limiar correspondente a **70% do máximo da Zona 1** é **36543.07658**.

O DataFrame `df_alto` reúne todos os registros em que o consumo da Zona 1 é superior a esse valor.


## 5. Contar os registros e calcular o percentual

In [6]:
qtd_alto = len(df_alto)
perc_alto = qtd_alto / len(dados) * 100

print(f"Total de registros do dataset: {len(dados)}")
print(f"Registros acima de 70% do máximo: {qtd_alto}")
print(f"Percentual: {perc_alto:.2f}%")


Total de registros do dataset: 52416
Registros acima de 70% do máximo: 14718
Percentual: 28.08%


### Resultado

Foram encontrados **14718 registros** com consumo da Zona 1 acima de 70% do valor máximo.

Esses registros correspondem a **28.08%** dos **52416 registros** do dataset.


## 6. Calcular a temperatura média e criar um segundo DataFrame

O segundo DataFrame utiliza simultaneamente as condições:

1. consumo da Zona 1 acima de 70% do máximo;
2. temperatura acima da média do dataset.


In [7]:
temp_media = dados["Temperature"].mean()

df_alto_temp = dados[
    (dados["Consumo_Zona_1"] > limiar_70) &
    (dados["Temperature"] > temp_media)
].copy()

qtd_alto_temp = len(df_alto_temp)
perc_alto_temp = qtd_alto_temp / len(dados) * 100

print(f"Temperatura média: {temp_media:.2f}")
print(f"Registros que atendem às duas condições: {qtd_alto_temp}")
print(f"Percentual sobre o dataset: {perc_alto_temp:.2f}%")

display(df_alto_temp.head())


Temperatura média: 18.81
Registros que atendem às duas condições: 9917
Percentual sobre o dataset: 18.92%


,DateTime,Temperature,Humidity,Wind Speed,general diffuse flows,diffuse flows,Consumo_Zona_1,Consumo_Zona_2,Consumo_Zona_3
8606,3/1/2017 18:20,19.09,63.92,0.080,78.00,88.80,38598.12766,23195.12195,22552.25806
8607,3/1/2017 18:30,18.87,64.25,0.082,57.07,64.95,41521.02128,25119.51220,23771.61290
8750,3/2/2017 18:20,18.95,69.45,4.915,65.21,63.51,39266.04255,23121.95122,23098.06452
8893,3/3/2017 18:10,19.76,56.90,4.915,119.40,131.40,37421.61702,21662.19512,20409.67742
8894,3/3/2017 18:20,19.76,56.33,4.914,109.30,121.60,40258.72340,23571.95122,22912.25806


### Resultado

A temperatura média registrada é de aproximadamente **18.81 °C**.

Ao exigir simultaneamente consumo elevado e temperatura acima da média, restaram **9917 registros**, equivalentes a **18.92%** do dataset.


## 7. Comparar os dois conjuntos

In [8]:
reducao = qtd_alto - qtd_alto_temp
reducao_relativa = reducao / qtd_alto * 100

comparacao = pd.DataFrame({
    "Conjunto": [
        "Consumo acima de 70% do máximo",
        "Consumo acima de 70% + temperatura acima da média"
    ],
    "Quantidade": [qtd_alto, qtd_alto_temp],
    "Percentual_do_dataset": [perc_alto, perc_alto_temp]
})

display(
    comparacao.style.format({
        "Percentual_do_dataset": "{:.2f}%"
    })
)

print(f"Redução na quantidade de registros: {reducao}")
print(f"Redução em relação ao primeiro DataFrame: {reducao_relativa:.2f}%")


,Conjunto,Quantidade,Percentual_do_dataset
0,Consumo acima de 70% do máximo,14718,28.08%
1,Consumo acima de 70% + temperatura acima da média,9917,18.92%


Redução na quantidade de registros: 4801
Redução em relação ao primeiro DataFrame: 32.62%


### Interpretação

O primeiro DataFrame contém **14718 registros (28.08%)** de alto consumo da Zona 1.

Após incluir a condição de **temperatura acima da média**, o segundo DataFrame passou a conter **9917 registros (18.92%)**.

Portanto, houve uma redução de **4801 registros**, o que corresponde a aproximadamente **32.62%** dos registros que inicialmente apresentavam consumo elevado.

Isso ocorre porque o segundo filtro é mais restritivo: para permanecer no novo conjunto, o registro precisa atender **às duas condições ao mesmo tempo**. Assim, parte dos momentos de alto consumo ocorreu com temperatura igual ou abaixo da média e foi excluída do segundo DataFrame.

### Conclusão

A **Zona 1** foi a zona com o maior pico de consumo, atingindo **52204.39512**. O limite de 70% desse pico foi **36543.07658**. Foram identificados **14718 registros (28.08%)** acima desse limite.

Quando também foi exigida temperatura superior à média de **18.81 °C**, a quantidade caiu para **9917 registros (18.92%)**. Portanto, a inclusão da condição ambiental reduziu o conjunto analisado, mostrando que nem todos os períodos de maior demanda da Zona 1 aconteceram sob temperaturas acima da média.


- Máximo Zona 1: 52.204,39512
- Máximo Zona 2: 37.408,86076
- Máximo Zona 3: 47.598,32636
- Maior pico: Zona 1
- 70% do máximo da Zona 1: 36.543,07658
- Registros acima desse limite: 14.718
- Percentual: 28,08%
- Temperatura média: 18,81 °C
- Registros com consumo alto + temperatura acima da média: 9.917
- Percentual: 18,92%
- Redução após incluir a condição ambiental: 4.801 registros, aproximadamente 32,62% do primeiro conjunto.